ufficio_soci/elenchi.py

```diff
@@ -677,13 +677,22 @@ def risultati(self):
             pk__in=tesserini_richiesti.values_list('id', flat=True)
 
         ).annotate(
-                appartenenza_tipo=F('appartenenze__membro'),
-                appartenenza_inizio=F('appartenenze__inizio'),
-                appartenenza_sede=F('appartenenze__sede'),
+            appartenenza_tipo=F('appartenenze__membro'),
+            appartenenza_inizio=F('appartenenze__inizio'),
+            appartenenza_sede=F('appartenenze__sede'),
         ).prefetch_related(
             'appartenenze', 'appartenenze__sede',
             'utenza', 'numeri_telefono'
         ).distinct('cognome', 'nome', 'codice_fiscale')
+        if tesserini_da_richiedere:
+            id_tesserini_da_richiedere = [i.id for i in tesserini_da_richiedere]
+            tesserini_rifiutati = Tesserino.objects.filter(persona_id__in=id_tesserini_da_richiedere,
+                                                    stato_richiesta=Tesserino.RIFIUTATO)
+            id_tesserini_rifiutati = tesserini_rifiutati.values_list('persona_id', flat=True).distinct()
+            if id_tesserini_rifiutati:
+                # persone con tesserino non rifiutato
+                tesserini_da_richiedere = tesserini_da_richiedere.exclude(pk__in=id_tesserini_rifiutati)
+        return tesserini_da_richiedere
```

Comment:
Non capisco questo blocco di codice, siamo sicuri che non sia possibile fare tutto semplicemente aggiungendo una riga in `.filter()` o `.exclude()`? 

Inoltre, questo blocco e' sfortunatamente veramente inefficiente, al punto da poter uccidere un nostro server applicativo (timeout). In particolare:

- Il metodo `risultati()` dovrebbe ritorare un QuerySet **non eseguito** (*unevaluated*). Queste query sono estremamente costose (possono impiegarchi decine di secondi, con database di dimensioni come il nostro). Ritornare un QuerySet **non eseguito** ci permette, ad esempio, di chiamare `.count()` sul QuerySet in modo molto efficiente, senza scaricare l'elenco (ma usando `COUNT()` di SQL dietro le quinte).
  Ti consiglio di leggere [QuerySets: When QuerySets are evaluated](https://docs.djangoproject.com/en/2.0/ref/models/querysets/#when-querysets-are-evaluated) dalla documentazione di Django. In particolare:
  - Utilizzare un queryset in una condizione (`if tesserini_da_richiedere`) ne forza l'esecuzione. Peggio ancora, senza usare `.exist()`, forza lo scaricamento di *tutti i risultati*, non paginati, in memoria! Vedi documentazione di [QuerySet.exist](https://docs.djangoproject.com/en/2.0/ref/models/querysets/#django.db.models.query.QuerySet.exists).
  - Utilizzare un generatore e/o una lista come argomento in `.filter()` causa l'esecuzioni di query multiple, impedendo all'ORM di Django di fare JOIN e/o ottimizzazioni dietro le quinte. Ricorda che un QuerySet puo' essere usato all'interno di un `__in` direttamente. In particolare,
    ```python
    id_tesserini_da_richiedere = [i.id for i in tesserini_da_richiedere]
    tesserini_rifiutati = Tesserino.objects.filter(persona_id__in=id_tesserini_da_richiedere,
    ```
    Puo' essere sostituito dal piu' semplice, e molto piu' efficiente, equivalente:
    ```python
    tesserini_rifiutati = Tesserino.objects.filter(persona__in=tesserini_da_richiedere, 
    ```
    Vedi la [documentazione delll'operatore `in`](https://docs.djangoproject.com/en/2.0/ref/models/querysets/#in) per maggiori informazioni.